operation/web.go

```diff
@@ -0,0 +1,118 @@
+package operation
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const opidMetadataKey = "opid"
+
+// UnaryClientInterceptor adds the operation id from the current context (if any) to the
+// outgoing unary RPC metadata.
+func UnaryClientInterceptor(
+	ctx context.Context,
+	method string,
+	req, reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	if op := Get(ctx); op != nil && op.ID.String() != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, opidMetadataKey, op.ID.String())
+	}
+	return invoker(ctx, method, req, reply, cc, opts...)
+}
+
+// StreamClientInterceptor adds the operation id from the current context (if any) to the
+// outgoing streaming RPC metadata.
+func StreamClientInterceptor(
+	ctx context.Context,
+	desc *grpc.StreamDesc,
+	cc *grpc.ClientConn,
+	method string,
+	streamer grpc.Streamer,
+	opts ...grpc.CallOption,
+) (grpc.ClientStream, error) {
+	if op := Get(ctx); op != nil && op.ID.String() != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, opidMetadataKey, op.ID.String())
+	}
+	return streamer(ctx, desc, cc, method, opts...)
+}
+
+// UnaryServerInterceptor creates a new operation in the current context before passing
+// it to the unary response handler. If the incoming RPC metadata contains an operation
+// ID, the new operation will have the same ID.
+func (m *Manager) UnaryServerInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	meta, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, errors.New("failed to pull metadata from context")
+	}
+	opid, err := GetOrCreateFromMetadata(meta)
+	if err != nil {
+		return nil, err
+	}
+	ctx, done := m.createWithID(ctx, opid, info.FullMethod, nil)
+	defer done()
```

Comment:
feels like we can refactor this out since the StreamServerInterceptor is very similar

Reply:
agreed, `GetOrCreateFromMeta` can probably house more of the common logic

Reply:
https://github.com/viamrobotics/rdk/pull/1501/commits/f921445bc0decb1c38e05c0dd764b076f5bcebc2